src/Hint/Monad.hs

```diff
@@ -238,8 +239,21 @@ monadSteps _ _ = []
 
 -- | Rewrite 'do ...; x <- return y; ...' as 'do ...; let x = y; ...'.
 monadLet :: [ExprLStmt GhcPs] -> [(ExprLStmt GhcPs, ExprLStmt GhcPs, Refactoring R.SrcSpan)]
-monadLet xs = mapMaybe mkLet xs
+monadLet xs = go mkLet xs
   where
+    go f = \case
+      [] -> []
+      -- If the do block is possibly an ApplicativeDo block, don't make a suggestion on the
+      -- second last statement.
+      [_, x] | mayBeApplicativeDo x -> go f [x]
+      x : xs -> maybe id (:) (f x) (go f xs)
+
+    -- A do block is possibly an ApplicativeDo block if the last statement is one of
+    -- 'pure ...' , 'pure $ ...', 'return ...' or 'return $ ...'.
+    mayBeApplicativeDo = \case
```

Comment:
I think this will have too many false positives. I would wager *most* do blocks end like this :/

Reply:
Agreed - but at the same time - the hint isn't the greatest in the world, so I'm less concerned than I would be otherwise.

Reply:
I think I've seen it a few times in student code, so I do consider it to be useful as a teaching tool at least.